### PR TITLE
Fix the cache and add tests

### PR DESCRIPTION
**Pull request by foursixnine**
Ensure that cache is working as expected, I.E Limits are being
followed correctly, that elements are deleted when they are removed
from the list of items, and that db file is updated accordingly.

Cache.db file was previously copied inside the pool directory where
the worker has no write permissions, this moves it to the proper value in
CACHEDIRECTORY where it can write.

### CODE REVIEW

**Comment by foursixnine** [comment]
@foursixnine _crosses fingers and prays to codecov's ${DEITIES} to allow the coverage to pass_

**Comment by okurz** [comment]
your prayers have been heard but maybe you want to ensure the full stack test produces reproducible and stable results in the future. We don't know if the gods' light will always shine on you ;-)

File: lib/OpenQA/Worker/Cache.pm [hunk 1]
```diff
@@ -151,18 +151,23 @@ sub get_asset {
...
     # currently only
     while (@{$cache->{$host}} > $limit) {
+        my $count = @{$cache->{$host}};
         my $asset = pop(@{$cache->{$host}});
-        unlink($asset) if $remove;
-        log_debug("Purged $asset due to $limit");
+        if (-e $asset) {
+            unlink($asset);
+            if ($@) {
+                log_error("Cannot purge $asset: $@");
```
Comment (okurz) [inline lib/OpenQA/Worker/Cache.pm]: Under which circumstances should this be acceptable?

**Comment by okurz** [comment]
… actually, where *is* the result from codecov?

**Comment by coolo** [comment]
https://codecov.io/gh/os-autoinst/openQA/commit/0bfdd37b0c4ad5b7ce11d38f41ea06ab1e20aac7 believes CI failed and so does not report. But the coverage is 87.23%

File: lib/OpenQA/Worker/Cache.pm (hunk 1 above)
Comment (coolo) [inline lib/OpenQA/Worker/Cache.pm]: possibly if 2 workers cleanup at the same time. The locking between 2 workers is rather non-existant

**Comment by coolo** [comment]
At least we have a test

**Comment by coolo** [comment]
the problem is still not gone though. The bigger problem is the non-existant locking between the workers. And so it can easily happen that we have 29 entries in the hash, but 187 files in the cache (as on openqaworker1.openqanet.opensuse.org)

**Comment by foursixnine** [comment]
Yeah this is the concurrency one i want to address

File: lib/OpenQA/Worker/Cache.pm [hunk 2]
```diff
@@ -56,6 +55,7 @@ sub set {
 sub init {
     my $class;
     ($host, $location) = @_;
+    $db_file = catdir($location, $db_file);
```
Comment (coolo) [inline lib/OpenQA/Worker/Cache.pm]: this breaks on re-init after api-failures and creates https://progress.opensuse.org/issues/18110

**Pull request merged**